### PR TITLE
Bugfixing and stabilization for sovereign chain

### PR DESCRIPTION
**Pull request by SebastianMarian**
## Reasoning behind the pull request

- Fixed signing and checking aggregate signature on consensus flow for sovereign chain
 
## Proposed changes
- 
- 
- 

## Testing procedure
- 
- 
- 

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

**Review by sasurobert — APPROVED** [review]

File: consensus/constants.go [hunk 1]
```diff
@@ -1,11 +1,11 @@
 package consensus
 
-// SubroundBlockType defines the types of subround block type
-type SubroundBlockType string
+// ConsensusModel defines the consensus models
+type ConsensusModel string
```
Comment (iulianpascalau) [inline consensus/constants.go:4]: :+1:

File: consensus/spos/bls/blsSubroundsFactory.go [hunk 2]
```diff
@@ -246,25 +273,53 @@ func (fct *factory) generateSignatureSubround() error {
...
+
+func (fct *factory) generateEndRoundSubround() error {
+	subroundEndRoundInstance, err := fct.generateEndRoundSubroundV1()
 	if err != nil {
 		return err
 	}
 
-	fct.worker.AddReceivedMessageCall(MtSignature, subroundSignatureObject.receivedSignature)
-	fct.consensusCore.Chronology().AddSubround(subroundSignatureObject)
+	switch fct.consensusModel {
```
Comment (iulianpascalau) [inline consensus/spos/bls/blsSubroundsFactory.go:297]: Can be made in another PR: instead of having multiple `switch` operations on each subround, we can have only one switch to create all the subrounds following the model 1 or model 2.

File: testscommon/components/components.go [hunk 3]
```diff
@@ -173,7 +173,7 @@ func GetConsensusArgs(shardCoordinator sharding.Coordinator) consensusComp.Conse
 		StatusComponents:     statusComponents,
 		StatusCoreComponents: GetStatusCoreComponents(),
 		ScheduledProcessor:   scheduledProcessor,
-		SubroundBlockType:    consensus.SubroundBlockTypeV1,
+		ConsensusModel:       consensus.ConsensusModelV1,
```
Comment (iulianpascalau) [inline testscommon/components/components.go:176]: For the next PR: can we have the consensus integration tests running on both variants? It will be way easier to debug if we need to in the future

**Review by iulianpascalau — DISMISSED** [review]

File: process/block/baseProcess.go [hunk 4]
```diff
@@ -341,6 +341,11 @@ func displayHeader(headerHandler data.HeaderHandler) []*display.LineData {
 	} else {
 		shardHeader, isShardHeader := headerHandler.(data.ShardHeaderHandler)
 		if isShardHeader {
+			_, isHeaderWithValidatorStats := headerHandler.(*block.HeaderWithValidatorStats)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/baseProcess.go:344]: +1

File: process/common.go [hunk 5]
```diff
@@ -26,6 +26,8 @@ import (
 
 var log = logger.GetOrCreate("process")
 
+var SovereignHeaderVersion = []byte("S1")
```
Comment (sasurobert) [inline process/common.go:29]: missing comment

File: process/common.go [hunk 6]
```diff
@@ -867,6 +875,11 @@ func UnmarshalShardHeader(marshalizer marshal.Marshalizer, hdrBuff []byte) (data
 		return hdr, nil
 	}
 
+	hdr, err = UnmarshalHeaderWithValidatorStats(marshalizer, hdrBuff)
```
Comment (sasurobert) [inline process/common.go:878]: why do we need a code like this ? can't we have a factory/interface for this unmarshal and 3 implementations. Why should we try unmarshalling for header v2 and only after doing for the one with validators stats ? It is double work on CPU. In case of importDB, it is triple work.

File: process/common.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/common.go:29]: Yep, added a TODO here about the approach that should be used whenever a new version of sovereign header will be created

Comment (SebastianMarian) [reply process/common.go:29]: Added

File: process/common.go (hunk 6 above)
Comment (SebastianMarian) [reply process/common.go:878]: Unfortunately could not be resolved for all the cases, as on header interceptor you will receive some bytes. All you know is that this should be a header. Then you should try to unmarshall in all these variants. To optimize this stuff at least we put the most probable implementation header, like v2 for now, as the first try.

File: process/block/baseProcess.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/block/baseProcess.go]: Done

File: process/common.go (hunk 6 above)
Comment (sasurobert) [reply process/common.go:882]: can't we have sovereignHeaderInterceptor and hearderV2Interceptor ? I think this part has to be refactored.

Comment (iulianpascalau) [reply process/common.go:882]: I can get the task of injecting an abstract factory that will get only the marshaled bytes and try to output a headerHandler. We will have 2 versions: one for normal chain operation, and one for the sovereign. Then we can try to optimize it further if we want. This will be a whole new PR.

**Review by sasurobert — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]
